Review Проектная работа E2E & Unit тестирование "Stellar Burger"

tryproxy opened the pull request:
1. Task: Подготовить и настроить конфиг для JEST и CYPRESS, покрыть тестами пользовательсике кейсы и redux store.
2. Screenshot: 
![cypress](https://github.com/user-attachments/assets/f776edeb-b8a5-4a3f-93cf-481d4c9fa295)
![jest](https://github.com/user-attachments/assets/04e0349c-3ba5-49b2-8da3-ad3bfa57ed31)
3. Done 05.08.2024

Реализовано:
- [x] Настроить среду тестирования;
- [X] Написать E2E тесты (Cypress);
- [X] Написать UNIT тесты (JEST);

tryproxy committed:
test: init cypress

tryproxy committed:
test(e2e): add mock data, add cypress config route;
add case adding items to burger constructor;
add case open/close modal to burger constructor;

tryproxy committed:
test(e2e): add case creating order;

tryproxy committed:
test: init jest;
test(unit): add rootReducer state

tryproxy committed:
test(unit): add burgerConstructorSlice;

tryproxy committed:
test(unit): add mocks directory, add tests directory;
test(unit): add cases for burgerIngredientSlice;

tryproxy committed:
test(unit): add cases for feedSlice;
test(unit): add cases for orderSlice;

tryproxy committed:
test(unit): add cases for profileSlice;
refactor: clean up;

gennady-bars commented on src/services/__tests__/burgerConstructorSlice.test.ts:
```diff
@@ -0,0 +1,119 @@
+} from '../__mocks__/ingredientsData';
+
+describe('Burger constructor state functionality', () => {
+  const initialState = {
```
`initialState ` не нужно создавать в тестах опять.  Нужно импортировать `initialState` из файла редьюсера. Там уже создан начальный стейт для каждого редьюсера

gennady-bars commented on cypress.config.ts:
```diff
@@ -0,0 +1,10 @@
+
+export default defineConfig({
+  e2e: {
+    baseUrl: 'http://localhost:4000',
```
* Отлично, что базовый урл вынесен в конфиг

gennady-bars commented on cypress/e2e/constructor.cy.tsx:
```diff
@@ -0,0 +1,101 @@
+
+  it('Open/close-btn modal popup with an ingredient', function () {
+    cy.get('[data-cy="burger-ingredient-card"]').contains('булк').click();
+    cy.get('[data-cy="burger-modal"]').should('exist');
```
`[data-cy="burger-modal"]` 8 раз повторяется 

Если **строка-селектор** повторяется более 2х раз в коде, значит, этот селектор нужно выносить в **константу**, чтобы можно было одним движением изменить его и не искать по всему коду дублирования этого селектора. 

**Можно лучше**

* Лучше использовать переменные **alias** в тестах, чтобы 1 раз найти элемент и обращаться к нему. Подробнее тут https://docs.cypress.io/api/commands/as#DOM-element 
* Попробуйте сделать команды `cypress`. Подробнее тут https://on.cypress.io/custom-commands

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная:


* Все юнит-тесты (38 шт) проходят успешно
* Все cypress-тесты проходят успешно
* Код тестов аккуратный и понятный
* Отлично, что базовый урл вынесен в конфиг


но есть некоторые недочеты:


* `initialState ` не нужно создавать в тестах опять.  Нужно импортировать `initialState` из файла редьюсера. Там уже создан начальный стейт для каждого редьюсера
* Если **строка-селектор** повторяется более 2х раз в коде, значит, этот селектор нужно выносить в **константу**, чтобы можно было одним движением изменить его и не искать по всему коду дублирования этого селектора. 

**Можно лучше**

* Лучше использовать переменные **alias** в тестах, чтобы 1 раз найти элемент и обращаться к нему. Подробнее тут https://docs.cypress.io/api/commands/as#DOM-element 
* Попробуйте сделать команды `cypress`. Подробнее тут https://on.cypress.io/custom-commands

tryproxy committed:
test(unit): add initialState export from slices;
test(e2e): replace repeated selectors w/ constants, add alias for modal;
test(e2e): add cypress custom command for popup;

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
